New option for zlib compression

valldrac opened the pull request:
Add a new option to enable compression using zlib library from the kernel.

I have adjusted the compression level to reduce memory footprint while providing a good speedup. It performs more than 4x in my tests with Android.

This PR also does some code refactor for future improvements and fixes minor bugs. It resolves #31.

jtsylve commented:
Great work.  I think before we merge this, we should extend the header format to tell whether the data is compressed or not, that way memory readers can read compressed images directly.

Have you experimented with compressing blocks larger than 4KiB?  I'm wondering if you might get better compression ratios that way (although it does increase the complexity a bit).

valldrac commented:
Thank you. I've pushed more commits but before digging into the details let me explain how it works.

It compresses the whole file just to transfer it faster over the network or sdcard, including the headers, padding, etc. Then it is very easy to pipe through a decompressor before analyzing it. I've updated the README to document that. I think is better this way because it's transparent for memory readers. Otherwise, you are forcing them to update to read compressed pages.

My first version flushes the compression buffer all the time. It is not exactly the same as compressing page by page, but it affects the compression rate a bit. So I rewrote it to manage the buffering as a stream, without flush.

I've introduced several enhancements and bug fixes. The digest output runs now. It was broken for TCP and file basically. The performance has also improved dramatically. Before it takes 15-20 minutes to acquire 3GB of RAM from android via ADB forwarding. Now it finishes in 3 minutes.

It's a huge PR with lots of changes. Please take your time to review and test it :)

kd8bny commented:
Super good stuff. This has been on my todo list for quite some time. Looking forward to running tests.

kd8bny commented:
Finished moving houses, will have some time to spin this back up

valldrac commented:
Thanks for your help @kd8bny. I have been using it for a while without any issue but in a very limited subset of devices. We need to test it thoroughly.

kd8bny commented:
@valldrac I think for the time being well move this to a beta branch as we need to get a path forward on including compression in the LiME header. This most likely break some readers (not sure) @jtsylve. Its working for me so far, I had to make an adjustment the the tcp accept method but all is good. Ill be conducting further testing on older KVERs. The compression is impressive, so much quicker

kd8bny commented:
@jtsylve @valldrac This is looking great in my tests. I am going to run some old 2.x tests

kd8bny commented:
We have some breakages on old 2.x kernels. Im going to merge this in a trunk branch so we can start some work.
